evm/src/executor/fork/backend.rs

```diff
@@ -235,100 +238,121 @@ where
-                ProviderRequest::BlockHash(fut) => {
-                    if let Poll::Ready((block_hash, number)) = fut.poll_unpin(cx) {
-                        let value = block_hash.unwrap_or_else(|_| {
-                            trace!("Failed to get block hash for {}", number);
+                    ProviderRequest::BlockHash(fut) => {
+                        if let Poll::Ready((block_hash, number)) = fut.poll_unpin(cx) {
+                            let value = block_hash.unwrap_or_else(|_| {
+                            trace!( target: "backendhandler", "Failed to get block hash for {}", number);
                             Default::default()
                         });
 
-                        // update the cache
-                        pin.db.block_hashes().write().insert(number, value);
+                            // update the cache
+                            pin.db.block_hashes().write().insert(number, value);
 
-                        // notify all listeners
-                        if let Some(listeners) = pin.block_requests.remove(&number) {
-                            listeners.into_iter().for_each(|l| {
-                                let _ = l.send(value);
-                            })
+                            // notify all listeners
+                            if let Some(listeners) = pin.block_requests.remove(&number) {
+                                listeners.into_iter().for_each(|l| {
+                                    let _ = l.send(value);
+                                })
+                            }
+                            continue
                         }
-                        continue
                     }
                 }
+                // not ready, insert and poll again
+                pin.pending_requests.push(request);
+            }
+
+            // If no new requests have been queued, break to
+            // be polled again later.
+            if pin.queued_requests.is_empty() {
+                break
             }
-            // not ready, insert and poll again
-            pin.pending_requests.push(request);
         }
 
-        // the handler is finished if the request channel was closed and all requests are processed
-        if pin.incoming.is_done() && pin.pending_requests.is_empty() {
-            if let Poll::Ready(Some(ack)) = Pin::new(&mut pin.shutdown).poll_next(cx) {
-                // effectively flushing the cached storage if any
+        if let Poll::Ready(Some(ack)) = Pin::new(&mut pin.shutdown).poll_next(cx) {
+            trace!(target: "backendhandler", "received shutdown signal");
+            if pin.incoming.is_done() {
+                // if request channel was closed, flush the cached storage to disk, if caching is
+                // enabled
                 pin.db.cache().flush();
-                // signaling back
-                let _ = ack.send(());
             }
+            // signaling back
+            let _ = ack.send(());
```

Comment:
> the problem i ran into is: if SharedBackend is cloned, then BackendHandler is not dopped (=> leaking memory)

this is indented behaviour because we want to be able to use multiple `SharedBackend`s connected to the same handler.

here is the chain of events that will cause the `BackendHandler` to exit.

1. last SharedBackend is dropped
2. `incoming` channel is closed, 
3. SharedBackend sends `shutdown`
4. `BackendHandler` receives shutdown 
5. `incoming.is_done == true` so cache is flushed
6. send ack back to  SharedBackend
7. SharedBackend dropped closing the shutdown channel
8. BackendHandler enters exit condition and returns Ready

But we could return ready if it was the last SharedBackend that sent the shutdown msg

Reply:
> 8. BackendHandler enters exit condition and returns Ready

after `BackendHandler::poll` returns `Poll::Pending`, there is nothing scheduled to wake it (call poll again) to finally exit & drop.

Reply:
that's interesting because I'd assume dropping the last shutdown sender(closing the channel) should awake the receiver's weaker.
just tested this and it's indeed inconsistent which is weird...
so the proper fix would be to exit after flushing the cache then

Reply:
> that's interesting because I'd assume dropping the last shutdown sender(closing the channel) should awake the receiver's waker.

yes, i didn't know how does it work either - it registers for 1 poll "iteration" & in this case there was no need to register again (value was available).
```rust
/// Attempt to pull out the next value of this stream, registering the
/// current task for wakeup if the value is not yet available
fn poll_next(...)
```

Reply:
btw, while we are at discussing backend - would you support this change? https://github.com/gakonst/foundry/pull/1003#discussion_r838435937

Reply:
supportive1